network/src/peers_registry.rs

```diff
@@ -112,235 +113,283 @@ impl PeerConnection {
+            max_inbound,
+            max_outbound,
             reserved_only,
-            deny_list,
         }
     }
 
     #[allow(clippy::needless_lifetimes)]
     #[inline]
     pub fn get_peer_id<'a>(&'a self, peer_index: PeerIndex) -> Option<&'a PeerId> {
-        self.peer_connections.get_peer_id(peer_index)
+        self.peers.get_peer_id(peer_index)
     }
 
-    // registry a new peer
-    #[allow(clippy::needless_pass_by_value)]
-    pub fn new_peer(&mut self, peer_id: PeerId, endpoint: Endpoint) -> Result<(), Error> {
-        if self.peer_connections.get(&peer_id).is_some() {
+    pub fn is_reserved(&self, peer_id: &PeerId) -> bool {
+        self.reserved_peers.contains(&peer_id)
+    }
+
+    pub fn accept_inbound_peer(&mut self, peer_id: PeerId, addr: Multiaddr) -> Result<(), Error> {
+        if self.peers.get(&peer_id).is_some() {
             return Ok(());
         }
-        let is_reserved = self.peer_store.read().is_reserved(&peer_id);
+        if !self.is_reserved(&peer_id) {
+            if self.reserved_only {
+                return Err(ErrorKind::InvalidNewPeer(format!(
+                    "We are in reserved_only mode, rejected non-reserved peer {:?}",
+                    peer_id
+                ))
+                .into());
+            }
+            let candidate_score = {
+                let peer_store = self.peer_store.read();
+                if peer_store.is_banned(&peer_id) {
+                    return Err(
+                        ErrorKind::InvalidNewPeer(format!("peer {:?} is denied", peer_id)).into(),
+                    );
+                }
+                peer_store.peer_score_or_default(&peer_id)
+            };
+            let connection_status = self.connection_status();
+            // check peers connection limitation
+            if connection_status.unreserved_inbound >= self.max_inbound
+                && !self.try_evict_inbound_peer(candidate_score)
+            {
+                return Err(ErrorKind::InvalidNewPeer(format!(
+                    "reach max inbound peers limitation, reject peer {:?}",
+                    peer_id
+                ))
+                .into());
+            }
+        }
+        self.new_peer(peer_id, addr, Endpoint::Listener);
+        Ok(())
+    }
+
+    fn try_evict_inbound_peer(&mut self, candidate_score: Score) -> bool {
+        let (peer_id, score) = {
+            let inbound_peers = self
+                .peers
+                .iter()
+                .filter(|(peer_id, peer)| peer.is_inbound() && !self.is_reserved(peer_id));
+            let candidate_peers = find_most_peers_in_same_network_group(inbound_peers);
+            let peer_store = self.peer_store.read();
+
+            let mut lowest_score = std::i32::MAX;
+            let mut low_score_peers = Vec::new();
+            for peer_id in candidate_peers {
+                if let Some(score) = peer_store.peer_score(peer_id) {
+                    if score > lowest_score {
```

Comment:
please use `min_by`

Reply:
What we want here is a random choice when there're multiple peers have the same score, `min_by` not guaranteed this.